File: frontend/src/types/fantasy.ts
import type { UserTitle } from './user'

export type FantasyLeague = {
  id: number
  tournamentId: number
  tournamentLink: string
  tournamentName: string | null
  started: boolean
  finished: boolean
  maxPlayers: number
  maxCost: number
}

export type FantasyPlayerRow = {
  id: number
  fantasyLeagueId: number
  tournamentPlayerId: number
  name: string | null
  link: string | null
  race: string | null
  cost: number
  pointsRo24: number | null
  pointsRo16: number | null
  pointsRo8: number | null
  pointsRo4: number | null
  pointsRo2: number | null
  pointsEarned: number
  defeated: boolean
  isWinner: boolean
  elo?: number
}

export type FantasyPreviewPlayer = {
  tournamentPlayerId: number
  name: string | null
  link: string | null
  race: string | null
  elo: number
  cost: number
}

export type FantasyTeamMemberRow = {
  fantasyPlayerId: number
  name: string | null
  link: string | null
  race: string | null
  cost: number
  pointsEarned: number
  defeated: boolean
  isWinner: boolean
  elo: number
}

export type FantasyTeamRow = {
  id: number
  fantasyLeagueId: number
  userId: number
  userAlias: string
  rank: number
  points: number
  cost: number
  members: FantasyTeamMemberRow[]
  titles?: UserTitle[]
}

export type TournamentSummary = {
  id: number
  link: string
  name: string | null
}

export const POINT_STAGES = ['Ro24', 'Ro16', 'Ro8', 'Ro4', 'Ro2'] as const
export type PointStage = (typeof POINT_STAGES)[number]

export function stagePoints(p: FantasyPlayerRow, stage: PointStage): number | null {
  switch (stage) {
    case 'Ro24':
      return p.pointsRo24
    case 'Ro16':
      return p.pointsRo16
    case 'Ro8':
      return p.pointsRo8
    case 'Ro4':
      return p.pointsRo4
    case 'Ro2':
      return p.pointsRo2
  }
}

export function stageReached(p: FantasyPlayerRow): number {
  let n = 0
  for (const stage of POINT_STAGES) {
    if (stagePoints(p, stage) != null) n++
    else break
  }
  return n
}

export function sortByElo<T extends { elo?: number | null }>(rows: T[]): T[] {
  return [...rows].sort((a, b) => (b.elo ?? 0) - (a.elo ?? 0))
}

File: backend/internal/model/fantasy.go
package model

// FantasyLeague is a fantasy competition bound to one tournament.
type FantasyLeague struct {
	ID             int64   `json:"id"`
	TournamentID   int64   `json:"tournamentId"`
	TournamentLink string  `json:"tournamentLink"`
	TournamentName *string `json:"tournamentName"`
	Started        bool    `json:"started"`
	Finished       bool    `json:"finished"`
	MaxPlayers     int     `json:"maxPlayers"`
	MaxCost        int     `json:"maxCost"`
}

// FantasyPlayerRow is a fantasy_player joined with tournament/player display fields.
type FantasyPlayerRow struct {
	ID                 int64    `json:"id"`
	FantasyLeagueID    int64    `json:"fantasyLeagueId"`
	TournamentPlayerID int64    `json:"tournamentPlayerId"`
	Name               *string  `json:"name"`
	Link               *string  `json:"link"`
	Race               *string  `json:"race"`
	Cost               int      `json:"cost"`
	PointsRo24         *int     `json:"pointsRo24"`
	PointsRo16         *int     `json:"pointsRo16"`
	PointsRo8          *int     `json:"pointsRo8"`
	PointsRo4          *int     `json:"pointsRo4"`
	PointsRo2          *int     `json:"pointsRo2"`
	PointsEarned       int      `json:"pointsEarned"`
	Defeated           bool     `json:"defeated"`
	IsWinner           bool     `json:"isWinner"`
	Elo                *float64 `json:"elo,omitempty"`
}

// FantasyPreviewPlayer is a tournament roster row with suggested fantasy cost.
type FantasyPreviewPlayer struct {
	TournamentPlayerID int64   `json:"tournamentPlayerId"`
	Name               *string `json:"name"`
	Link               *string `json:"link"`
	Race               *string `json:"race"`
	Elo                float64 `json:"elo"`
	Cost               int     `json:"cost"`
}

// FantasyTeamMemberRow is one roster slot on a fantasy team.
type FantasyTeamMemberRow struct {
	FantasyPlayerID int64   `json:"fantasyPlayerId"`
	Name            *string `json:"name"`
	Link            *string `json:"link"`
	Race            *string `json:"race"`
	Cost            int     `json:"cost"`
	PointsEarned    int     `json:"pointsEarned"`
	Defeated        bool    `json:"defeated"`
	IsWinner        bool    `json:"isWinner"`
	Elo             float64 `json:"elo"`
}

// FantasyTeamRow is a fantasy team with owner alias and members.
type FantasyTeamRow struct {
	ID              int64                  `json:"id"`
	FantasyLeagueID int64                  `json:"fantasyLeagueId"`
	UserID          int64                  `json:"userId"`
	UserAlias       string                 `json:"userAlias"`
	Rank            int                    `json:"rank"`
	Points          int                    `json:"points"`
	Cost            int                    `json:"cost"`
	Members         []FantasyTeamMemberRow `json:"members"`
	Titles          []UserTitle            `json:"titles"`
}

// FantasyPlayerCostOverride sets cost for a tournament player at create time.
type FantasyPlayerCostOverride struct {
	TournamentPlayerID int64 `json:"tournamentPlayerId"`
	Cost               int   `json:"cost"`
}

// FantasyGroupPlayer is a tournament group member with fantasy cost (no points).
type FantasyGroupPlayer struct {
	FantasyPlayerID int64   `json:"fantasyPlayerId"`
	Name            *string `json:"name"`
	Link            *string `json:"link"`
	Race            *string `json:"race"`
	Cost            int     `json:"cost"`
	Excluded        bool    `json:"excluded"`
	// IsGroupWinner marks standings advancement / group top (not fantasy champion).
	IsGroupWinner bool `json:"isGroupWinner"`
}

// FantasyGroup is a tournament group with fantasy costs for drafting.
type FantasyGroup struct {
	ID        int64                `json:"id"`
	Name      string               `json:"name"`
	Phase     string               `json:"phase"`
	SortOrder int                  `json:"sortOrder"`
	Players   []FantasyGroupPlayer `json:"players"`
}

// FantasyMatchBoard is groups + results for the Results tab / today panel.
type FantasyMatchBoard struct {
	Groups  []FantasyGroup `json:"groups"`
	Results []Result       `json:"results"`
	Today   string         `json:"today"` // YYYY-MM-DD UTC
}
